From 4e6b37f8a96f4c1c86b9d9e53c3c9d3f0ea24c5c Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Wed, 11 Jul 2018 13:18:29 +0300
Subject: [PATCH] board: habana: rework remote_load_firmware script

Current script assumes that there is a ready setup of the repo stuff
and also relies on some additional aliases that should be in place.
This is prone to repo corruptions and can't be ported simply to new machines.

Rework the remote_load_firmware script to not assume anything, but setup
all the needed prerequisites by itself.
Since the above might take a lot of time, move all the stuff to tmpfs.
This also makes the process more robust and immune to repo corruptions.

Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/board/habanalabs/common/remote_load_firmware b/board/habanalabs/common/remote_load_firmware
index 3d56689..64c4505 100755
--- a/board/habanalabs/common/remote_load_firmware
+++ b/board/habanalabs/common/remote_load_firmware
@@ -4,26 +4,70 @@
 success_str="1000"
 result=~/remote_load_firmware.result
 firm_dir=/lib/firmware/habanalabs/dali
-rm -f $result
+stack='~/trees/npu-stack'
+
+function remote_create_repo_sync() {
+	# prepare the repo
+	ssh -t $host "mkdir -p $stack; sudo mkdir -p $firm_dir;
+		sudo umount -lrf $stack; sleep 1;
+		sudo mount -t tmpfs -o uid=$UID,dirsync,mode=0755 tmpfs $stack;
+		git config --global user.name Jenkins;
+		git config --global user.email jenkins@habana.ai;
+		git config --global pull.rebase true;
+		cd $stack;
+		repo init -u ssh://192.168.0.59:29418/software-repo;
+		repo sync;" | tee -a $result
+
+	local pipestat=${PIPESTATUS[0]}
+	echo ssh repo sync status: $pipestat
+	if [ $pipestat -ne 0 ]; then
+		ssh -t 'sudo umount -lrf $stack;'
+		return $pipestat;
+	fi
+
+	ssh -t $host "cp $stack/automation/habana_scripts/.bashrc ~/ &&
+		cp $stack/automation/habana_scripts/.bash_aliases ~/;" | tee -a $result
+	local pipestat=${PIPESTATUS[0]}
+	echo ssh cp status: $pipestat
+	if [ $pipestat -ne 0 ]; then
+		ssh -t "sudo umount -lrf $stack;"
+	fi
+
+	return $pipestat;
+}
+
+function remote_bi_habanalabs() {
+	# build and insmod the habanalabs driver
+	# this depends on successful copy of the .bashrc and .bash_aliases
+	ssh -t $host bash -ic 'build_and_insmod_habanalabs -p "cpu_enable=1 cpu_queues_enable=1"' | tee -a $result
+	local success=`grep -cE "^$success_str" $result`
+	[ $success -ge 1 ] && return 0;
+
+	return 1;
+}
+
+function remote_copy_firmware() {
+	# sync the latest built binaries
+	for ((j=0; j<3; j++)); do
+		sshpass -p123456 scp ${firm_dir}/* root@${host}:${firm_dir}/ || continue;
+		return 0;
+	done
+
+	echo "Failed copying the latest firmware binaries..."
+	return 1;
+}
 
 for ((i=100; i>0; i--)); do
 	echo $i
 	sleep 1;
 	ping -c 1 -W 1 $host
 	if [ $? -eq 0 ]; then
-		ssh -t $host bash -ic 'hl_repo_sync' | tee -a $result
-		pipestat=${PIPESTATUS[0]}
-		echo ssh repo sync status: $pipestat
-		[ $pipestat -ne 0 ] && continue;
-
-		sshpass -p123456 scp ${firm_dir}/* root@${host}:${firm_dir}/
-
-		ssh -t $host bash -ic 'bi_habanalabs' | tee -a $result
-		success=`grep -cE "^$success_str" $result`
-		if [ $success -ge 1 ]; then
-			echo done
-			exit 0;
-		fi
+		rm -f $result
+		remote_create_repo_sync || continue;
+		remote_copy_firmware || break;
+		remote_bi_habanalabs || continue;
+		echo done;
+		exit 0;
 	fi
 done
 
